src/daemon/tabswitcher.vala

```diff
@@ -191,43 +315,80 @@ namespace Budgie {
-			if (window == null) {
-				return;
-			}
-			var desktop = this.app_system.query_window(window);
+		private void remove_window(TabSwitcherWidget? widget) {
+			if (widget == null) return;
+			ids.remove(widget.id);
+			window_box.remove(widget);
+			unowned List<string> entries = recency.find_custom(widget.id, strcmp);
+			recency.remove_link(entries);
+		}
 
-			var child = new TabSwitcherWidget(window, desktop, usertime);
-			xids.insert(xid, child);
+		private void set_window_as_activated(libxfce4windowing.Window window) {
+			string id = window.get_id().to_string();
+			unowned List<string> entries = recency.find_custom(id, strcmp);
+			recency.remove_link(entries);
+			recency.prepend(id);
+			window_box.invalidate_sort(); // Re-sort
+		}
 
-			/* Adjust to a maximum of 8 children per row */
-			var n_kids = xids.size();
-			if (n_kids < 8) {
-				window_box.set_max_children_per_line(n_kids);
-			} else {
-				window_box.set_max_children_per_line(8);
-			}
+		private void update_show_all_windows() {
+			if (settings == null) return;
+			show_all_windows = settings.get_boolean(SHOW_ALL_WINDOWS_KEY);
+			window_box.invalidate_filter(); // Re-filter
+		}
 
-			window_box.insert(child, -1);
-			queue_resize();
-			window_box.show_all();
+		private bool window_on_active_workspace(libxfce4windowing.Window window) {
+			unowned libxfce4windowing.Workspace? win_workspace = window.get_workspace(); // Get workspace
+			if (active_workspace == null || win_workspace == null) return true;
+			return win_workspace.get_id() == active_workspace.get_id();
 		}
 
 		/* Switch focus to the item with the xid */
-		public void focus_item(uint32 xid) {
-			/* Get the index of the xid it will be the same as the one the widget has */
-			TabSwitcherWidget? widget = xids.lookup(xid);
-			if (widget == null) {
-				return;
+		public void focus_item(bool backwards) {
+			unowned libxfce4windowing.Window? active_window = xfce_screen.get_active_window();
+			TabSwitcherWidget? widget = active_window != null ? ids.get(active_window.get_id().to_string()) : null;
+
+			// Visible, each input should cycle to previous / next
+			if (visible) {
+				widget = window_box.get_selected_children().nth_data(0).get_child() as TabSwitcherWidget;
+				if (widget != null) active_window = widget.window;
+			} else if (!visible && widget != null) {
+				window_box.select_child(widget.get_parent() as FlowBoxChild);
 			}
-			window_title.set_text(widget.title);
-			window_box.select_child(widget.get_parent() as Gtk.FlowBoxChild);
+
+			FlowBoxChild? new_child = null;
+			TabSwitcherWidget? new_widget = null;
+
+			var len = recency.length();
+			int64 id_pos = (widget != null) ? get_position_in_recency(widget.id).clamp(0, len - 1) : 0;
+
+			for (var i = 0; i < len; i++) {
+				uint64 new_id_pos = 0;
+
+				if (backwards) {
+					new_id_pos = (id_pos == 0 ? len : id_pos) - (i+1);
+				} else {
+					new_id_pos = id_pos == len - 1 ? 0 : id_pos + (i+1);
+				}
```

Comment:
Could some comments be added explaining what this is? Maybe I'm just not awake yet, but I have no idea what's happening here.

Reply:
Should be sorted as of cdb3a5e